### PR TITLE
Generator Resume Arguments

### PR DESCRIPTION
**Pull request by jonas-schievink**
cc https://github.com/rust-lang/rust/issues/43122 and https://github.com/rust-lang/rust/issues/56974

Blockers:
* [x] Fix miscompilation when resume argument is live across a yield point (https://github.com/rust-lang/rust/pull/68524#issuecomment-578459069)
* [x] Fix 10% compile time regression in `await-call-tree` benchmarks (https://github.com/rust-lang/rust/pull/68524#issuecomment-578487162)
  * [x] Fix remaining 1-3% regression (https://github.com/rust-lang/rust/pull/68524#issuecomment-579566255) - resolved (https://github.com/rust-lang/rust/pull/68524#issuecomment-581144901)
* [x] Make dropck rules account for resume arguments (https://github.com/rust-lang/rust/pull/68524#issuecomment-578541137)

Follow-up work:
* Change async/await desugaring to make use of this feature
* Rewrite [`box_region.rs`](https://github.com/rust-lang/rust/blob/3d8778d767f0dde6fe2bc9459f21ead8e124d8cb/src/librustc_data_structures/box_region.rs) to use resume arguments (this shows up in profiles too)

### CODE REVIEW

**Comment by petrochenkov** [comment]
Glad to see this implemented, but it is not my area of expertise.

r? @Zoxc for review or further reassignment.

File: src/libcore/ops/generator.rs [hunk 1]
```diff
@@ -84,6 +84,14 @@ pub trait Generator {
     /// represents a completed future.
     type Return;
 
+    /// The type of argument this generator takes when resuming.
+    ///
+    /// This associated type corresponds to the argument the generator takes,
+    /// and to the type of the `yield` expression inside the generator.
+    // NB: The order of the assoc. types is important here!
+    #[cfg(not(bootstrap))]
+    type Resume;
```
Comment (cramertj) [inline src/libcore/ops/generator.rs]: Note that having this be an associated type rather than a type parameter to the trait adds significant limitations, probably the most important of which is that it cannot contain any lifetimes that aren't `'static` or built into the type of the generator. I'd expect to see it as a type parameter, or at least a GAT in the final generator design.

I don't think this is worth blocking this PR on, as it's useful to have an implementation here, especially for unblocking no-std futures, but it does seem worthy of a comment.

Comment (jonas-schievink) [reply src/libcore/ops/generator.rs]: Yeah, that's the approach https://github.com/cavedweller/rust/commit/11c94220a62cd937090e9e6a04e50bb5f434ba6e took, but when I noticed that I was already ~80% finished. I'll open up a follow-up issue for that once this is merged.

File: src/librustc_mir/dataflow/move_paths/builder.rs [hunk 2]
```diff
@@ -380,7 +380,9 @@ impl<'b, 'a, 'tcx> Gatherer<'b, 'a, 'tcx> {
                 self.gather_operand(discr);
             }
 
-            TerminatorKind::Yield { ref value, .. } => {
+            TerminatorKind::Yield { ref value, resume_arg: ref place, .. } => {
+                self.create_move_path(place);
+                self.gather_init(place.as_ref(), InitKind::Deep);
```
Comment (Zoxc) [inline src/librustc_mir/dataflow/move_paths/builder.rs]: I noticed that the order of `gather_init` and `gather_operand` is swapped compared to `DropAndReplace`, but I'm not sure if it matters.

**Comment by Zoxc** [comment]
I think I'd prefer to make this PR make the argument be a type parameter instead of a associated type before landing to avoid some unnecessary churn in the compiler and for users of generators.

I also don't know how correct this is with regards to drops and references, but those concerns could be delayed by just adding a `'static + Copy` bound to the generator argument in typeck. That would be good enough for async/await.

**Comment by bjorn3** [comment]
I think an assocated type makes more sense, as it is very unlikely to ever be overloaded for different resume types.

**Comment by matthewjasper** [comment]
Async generators (should) implement `for<'a, 'b> Geneator<&'a Context<'b>, Yield = ()>`, which needs a type parameter.

File: src/libcore/ops/generator.rs (hunk 1 above)
Comment (jonas-schievink) [reply src/libcore/ops/generator.rs]: Turns out this isn't terribly complicated, so I implemented it now.

**Comment by jonas-schievink** [comment]
> I think I'd prefer to make this PR make the argument be a type parameter instead of a associated type before landing to avoid some unnecessary churn in the compiler and for users of generators.

Done!

> I also don't know how correct this is with regards to drops and references, but those concerns could be delayed by just adding a `'static + Copy` bound to the generator argument in typeck. That would be good enough for async/await.

True, but I'd prefer to not limit this feature to what async/await needs (that always leaves a sour taste in my mouth). I'm willing to add more tests, if you (or anyone else) have any suggestions for what needs to be tested here. The current tests are fairly minimal, but I did add some drop tests.

**Comment by jonas-schievink** [comment]
There's still a lingering miscompilation somewhere:

```rust
    let mut g = |mut d| {
        d = yield;
    };
```

This ends up dropping `d` after the yield, but without storing anything in its field in the generator state.

Could this be because the transform assumes there will be at least one `Assign` statement to a local before it needs to be saved? For locals inside the generator that makes sense, but local `_2` is passed in from the outside.

**Comment by jonas-schievink** [comment]
Let's see if the `yield` lowering changes affect compile times.

@bors try @rust-timer queue

**Comment by Zoxc** [comment]
@jonas-schievink Could you post the pre and post transform MIR of your example? The liveness result would be useful too.

**Comment by jonas-schievink** [comment]
This should show there's no store to the generator field before the yield:

<details>
  <summary>`generator_resume` MIR</summary>

```rust
// MIR for `main::{{closure}}#0`
// source = MirSource { instance: Item(DefId(0:11 ~ main[317d]::main[0]::{{closure}}[0])), promoted: None }
// pass_name = generator_resume
// disambiguator = 0
// generator_layout = GeneratorLayout { field_tys: [Dropper], variant_fields: [[], [], [], [_0]], storage_conflicts: BitMatrix { num_rows: 1, num_columns: 1, words: [1], marker: PhantomData } }

fn  main::{{closure}}#0(_1: std::pin::Pin<&mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}]>, _2: Dropper) -> std::ops::GeneratorState<(), ()> {
    debug d => (((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}])) as variant#3).0: Dropper); // in scope 0 at main.rs:15:18: 15:23
    let mut _0: std::ops::GeneratorState<(), ()>; // return place in scope 0 at main.rs:15:17: 17:6
    let mut _3: Dropper;                 // in scope 0 at main.rs:16:13: 16:18
    let mut _4: ();                      // in scope 0 at main.rs:16:13: 16:18
    let mut _5: ();                      // in scope 0 at main.rs:15:25: 15:25
    let mut _6: isize;                   // in scope 0 at main.rs:15:17: 17:6

    bb0: {
        _6 = discriminant((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}]))); // bb0[0]: scope 0 at main.rs:15:17: 17:6
        switchInt(move _6) -> [0u32: bb1, 1u32: bb12, 2u32: bb13, 3u32: bb11, otherwise: bb14]; // bb0[1]: scope 0 at main.rs:15:17: 17:6
    }

    bb1: {
        StorageLive(_3);                 // bb1[0]: scope 0 at main.rs:16:13: 16:18
        StorageLive(_4);                 // bb1[1]: scope 0 at main.rs:16:13: 16:18
        _4 = ();                         // bb1[2]: scope 0 at main.rs:16:13: 16:18
        _0 = std::ops::GeneratorState::<(), ()>::Yielded(move _4,); // bb1[3]: scope 0 at main.rs:16:13: 16:18
        discriminant((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}]))) = 3; // bb1[4]: scope 0 at main.rs:16:13: 16:18
        return;                          // bb1[5]: scope 0 at main.rs:16:13: 16:18
    }

    bb2 (cleanup): {
        discriminant((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}]))) = 2; // bb2[0]: scope 0 at main.rs:15:17: 17:6
        resume;                          // bb2[1]: scope 0 at main.rs:15:17: 17:6
    }

    bb3: {
        StorageDead(_4);                 // bb3[0]: scope 0 at main.rs:16:17: 16:18
        drop((((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}])) as variant#3).0: Dropper)) -> [return: bb8, unwind: bb7]; // bb3[1]: scope 0 at main.rs:16:9: 16:10
    }

    bb4 (cleanup): {
        goto -> bb9;                     // bb4[0]: scope 0 at main.rs:17:5: 17:6
    }

    bb5: {
        goto -> bb10;                    // bb5[0]: scope 0 at main.rs:17:5: 17:6
    }

    bb6: {
        _0 = std::ops::GeneratorState::<(), ()>::Complete(move _5,); // bb6[0]: scope 0 at main.rs:17:6: 17:6
        discriminant((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}]))) = 1; // bb6[1]: scope 0 at main.rs:17:6: 17:6
        return;                          // bb6[2]: scope 0 at main.rs:17:6: 17:6
    }

    bb7 (cleanup): {
        (((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}])) as variant#3).0: Dropper) = move _3; // bb7[0]: scope 0 at main.rs:16:9: 16:10
        StorageDead(_3);                 // bb7[1]: scope 0 at main.rs:16:17: 16:18
        drop((((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}])) as variant#3).0: Dropper)) -> bb4; // bb7[2]: scope 0 at main.rs:17:5: 17:6
    }

    bb8: {
        (((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}])) as variant#3).0: Dropper) = move _3; // bb8[0]: scope 0 at main.rs:16:9: 16:10
        StorageDead(_3);                 // bb8[1]: scope 0 at main.rs:16:17: 16:18
        _5 = ();                         // bb8[2]: scope 0 at main.rs:15:25: 17:6
        drop((((*(_1.0: &mut [generator@main.rs:15:17: 17:6 {Dropper, (), Dropper}])) as variant#3).0: Dropper)) -> [return: bb5, unwind: bb2]; // bb8[3]: scope 0 at main.rs:17:5: 17:6
    }

    bb9 (cleanup): {
        goto -> bb2;                     // bb9[0]: scope 0 at main.rs:17:5: 17:6
    }

    bb10: {
        goto -> bb6;                     // bb10[0]: scope 0 at main.rs:17:5: 17:6
    }

    bb11: {
        StorageLive(_3);                 // bb11[0]: scope 0 at main.rs:15:17: 17:6
        StorageLive(_4);                 // bb11[1]: scope 0 at main.rs:15:17: 17:6
        _3 = move _2;                    // bb11[2]: scope 0 at main.rs:15:17: 17:6
        goto -> bb3;                     // bb11[3]: scope 0 at main.rs:15:17: 17:6
    }

    bb12: {
        assert(const false, "generator resumed after completion") -> bb12; // bb12[0]: scope 0 at main.rs:15:17: 17:6
                                         // ty::Const
                                         // + ty: bool
                                         // + val: Value(Scalar(0x00))
                                         // mir::Constant
                                         // + span: main.rs:15:17: 17:6
                                         // + literal: Const { ty: bool, val: Value(Scalar(0x00)) }
    }

    bb13: {
        assert(const false, "generator resumed after panicking") -> bb13; // bb13[0]: scope 0 at main.rs:15:17: 17:6
                                         // ty::Const
                                         // + ty: bool
                                         // + val: Value(Scalar(0x00))
                                         // mir::Constant
                                         // + span: main.rs:15:17: 17:6
                                         // + literal: Const { ty: bool, val: Value(Scalar(0x00)) }
    }

    bb14: {
        unreachable;                     // bb14[0]: scope 0 at main.rs:15:17: 17:6
    }
}
```

</details>

I'm more worried about the benchmark results though. `normalize_ty_after_erasing_regions` takes 10% longer but is being invoked the same number of times as before.

**Comment by jonas-schievink** [comment]
The performance regression might be due to a more complex generator type being passed to the query. The logs show this:

> `fn([static generator@src/lib.rs:6:30: 10:10 {impl std::future::Future, impl std::future::Future, (), (), impl std::future::Future, impl std::future::Future, (), (), impl std::future::Future, impl std::future::Future, (), ()}]) -> impl std::future::Future {std::future::from_generator::<[static generator@src/lib.rs:6:30: 10:10 {impl std::future::Future, impl std::future::Future, (), (), impl std::future::Future, impl std::future::Future, (), (), impl std::future::Future, impl std::future::Future, (), ()}]>}`

All these `()` are likely resume arguments returned by `yield`s. I'm not sure why they now show up in the witness but before didn't (since `yield` did also have type `()` before).

Also, the dropck code seems to assume that only upvars contribute to dropck constraints, which is no longer true. We probably need to consider the entire interior there, no?

https://github.com/rust-lang/rust/blob/698fcd38fa9548e64a2092ff48c9d15ceb57d40c/src/librustc_traits/dropck_outlives.rs#L209-L239

**Comment by jonas-schievink** [comment]
> All these `()` are likely resume arguments returned by `yield`s. I'm not sure why they now show up in the witness but before didn't (since `yield` did also have type `()` before).

Hmm, nope. They're also there on rustc built from master :/ Really not sure what's going on then.

File: src/librustc_mir/transform/generator.rs [hunk 3]
```diff
@@ -1084,6 +1087,16 @@ where
...
 
+                // Move the resume argument to the destination place of the `Yield` terminator
+                let resume_arg = Local::new(2); // 0 = return, 1 = self
+                statements.push(Statement {
+                    source_info,
+                    kind: StatementKind::Assign(box (
+                        point.resume_arg,
+                        Rvalue::Use(Operand::Move(resume_arg.into())),
+                    )),
+                });
```
Comment (tmandry) [inline src/librustc_mir/transform/generator.rs]: This shouldn't happen for drop cases.

Comment (tmandry) [inline src/librustc_mir/transform/generator.rs]: This shouldn't happen for drop cases.

**Comment by jonas-schievink** [comment]
Added a fast path.

@bors try @rust-timer queue

**Comment by jonas-schievink** [comment]
Hmm, that didn't work. @bors try @rust-timer queue

File: src/librustc_traits/dropck_outlives.rs [hunk 4]
```diff
@@ -227,15 +227,16 @@ fn dtorck_constraint_for_ty<'tcx>(
...
+            // derived from lifetimes attached to the upvars and resume
+            // argument, and we *do* incorporate those here.
 
             constraints.outlives.extend(
                 substs
                     .as_generator()
                     .upvar_tys(def_id, tcx)
                     .map(|t| -> ty::subst::GenericArg<'tcx> { t.into() }),
             );
+            constraints.outlives.push(substs.as_generator().resume_ty(def_id, tcx).into());
```
Comment (Zoxc) [inline src/librustc_traits/dropck_outlives.rs]: This is wrong. The resume argument type is not a type stored in the generator in the same way that argument types for closures are not stored in the closures themselves.

File: src/librustc/traits/query/normalize.rs [hunk 5]
```diff
@@ -81,6 +81,10 @@ impl<'cx, 'tcx> TypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
     }
 
     fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
+        if !ty.has_projections() {
```
Comment (Zoxc) [inline src/librustc/traits/query/normalize.rs]: You should split this out to a separate PR since any performance changes from this PR will be entirely hidden by this performance improvement.

File: src/librustc_traits/dropck_outlives.rs (hunk 4 above)
Comment (jonas-schievink) [reply src/librustc_traits/dropck_outlives.rs]: Wouldn't that mean we'd have to consider the entire generator witness instead (since any type contained within may be stored in the generator across a `yield`)? AFAICT this is only dangerous if the types saved across a `yield` contain anything passed in from the outside (so upvars or the resume argument), right?

For closures, the only storage they can possibly have *is* their upvars, so there it works out.

File: src/librustc/traits/query/normalize.rs (hunk 5 above)
Comment (jonas-schievink) [reply src/librustc/traits/query/normalize.rs]: Done https://github.com/rust-lang/rust/pull/68606

**Comment by jonas-schievink** [comment]
Now with #68606 in master, let's see if there's any noticeable performance regression left.

@bors try @rust-timer queue

**Comment by jonas-schievink** [comment]
Still a 1-3% regression in the await-call-tree benchmarks

**Comment by jonas-schievink** [comment]
Temporarily cherry-picked https://github.com/rust-lang/rust/pull/68672, let's see if that fixes the remaining regression.

@bors try @rust-timer queue

**Comment by jonas-schievink** [comment]
@bors try-

This doesn't really work unless that PR is in master I think

**Comment by jonas-schievink** [comment]
@bors try @rust-timer queue

One more run after https://github.com/rust-lang/rust/pull/68672 has landed

**Comment by jonas-schievink** [comment]
Aaaand it's gone :tada:

**Comment by jonas-schievink** [comment]
I've pushed an attempt at fixing the miscompilation I found above, but it's still there (although the generated MIR does look better now). I don't immediately see what's wrong with this MIR though:


<details>
  <summary>`generator_resume` MIR</summary>

```rust
// MIR for `main::{{closure}}#0`
// source = MirSource { instance: Item(DefId(0:14 ~ resume_live_across_yield[317d]::main[0]::{{closure}}[0])), promoted: None }
// pass_name = generator_resume
// disambiguator = 0
// generator_layout = GeneratorLayout { field_tys: [Dropper], variant_fields: [[], [], [], [_0]], storage_conflicts: BitMatrix { num_rows: 1, num_columns: 1, words: [1], marker: PhantomData } }

fn  main::{{closure}}#0(_1: std::pin::Pin<&mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}]>, _2: Dropper) -> std::ops::GeneratorState<(), Dropper> {
    debug _d => (((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}])) as variant#3).0: Dropper); // in scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:18: 21:24
    let mut _0: std::ops::GeneratorState<(), Dropper>; // return place in scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
    let mut _3: Dropper;                 // in scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
    let mut _4: ();                      // in scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
    let mut _5: Dropper;                 // in scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:26: 21:26
    let mut _6: Dropper;                 // in scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:18: 21:24
    let mut _7: isize;                   // in scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6

    bb0: {
        _7 = discriminant((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}]))); // bb0[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
        switchInt(move _7) -> [0u32: bb1, 1u32: bb11, 2u32: bb12, 3u32: bb10, otherwise: bb13]; // bb0[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
    }

    bb1: {
        (((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}])) as variant#3).0: Dropper) = move _2; // bb1[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
        StorageLive(_3);                 // bb1[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
        StorageLive(_4);                 // bb1[2]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
        _4 = ();                         // bb1[3]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
        _0 = std::ops::GeneratorState::<(), Dropper>::Yielded(move _4,); // bb1[4]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
        discriminant((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}]))) = 3; // bb1[5]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
        return;                          // bb1[6]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:14: 22:19
    }

    bb2 (cleanup): {
        discriminant((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}]))) = 2; // bb2[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
        resume;                          // bb2[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
    }

    bb3: {
        StorageDead(_4);                 // bb3[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:18: 22:19
        drop((((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}])) as variant#3).0: Dropper)) -> [return: bb7, unwind: bb6]; // bb3[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:9: 22:11
    }

    bb4 (cleanup): {
        goto -> bb8;                     // bb4[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:5: 24:6
    }

    bb5: {
        _0 = std::ops::GeneratorState::<(), Dropper>::Complete(move _5,); // bb5[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:6: 24:6
        discriminant((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}]))) = 1; // bb5[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:6: 24:6
        return;                          // bb5[2]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:6: 24:6
    }

    bb6 (cleanup): {
        (((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}])) as variant#3).0: Dropper) = move _3; // bb6[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:9: 22:11
        StorageDead(_3);                 // bb6[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:18: 22:19
        drop((((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}])) as variant#3).0: Dropper)) -> bb4; // bb6[2]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:5: 24:6
    }

    bb7: {
        (((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}])) as variant#3).0: Dropper) = move _3; // bb7[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:9: 22:11
        StorageDead(_3);                 // bb7[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:22:18: 22:19
        _5 = move (((*(_1.0: &mut [generator@/home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6 {Dropper, ()}])) as variant#3).0: Dropper); // bb7[2]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:23:9: 23:11
        goto -> bb9;                     // bb7[3]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:5: 24:6
    }

    bb8 (cleanup): {
        goto -> bb2;                     // bb8[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:5: 24:6
    }

    bb9: {
        goto -> bb5;                     // bb9[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:24:5: 24:6
    }

    bb10: {
        StorageLive(_2);                 // bb10[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
        StorageLive(_3);                 // bb10[1]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
        StorageLive(_4);                 // bb10[2]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
        _3 = move _2;                    // bb10[3]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
        goto -> bb3;                     // bb10[4]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
    }

    bb11: {
        assert(const false, "generator resumed after completion") -> bb11; // bb11[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
                                         // ty::Const
                                         // + ty: bool
                                         // + val: Value(Scalar(0x00))
                                         // mir::Constant
                                         // + span: /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
                                         // + literal: Const { ty: bool, val: Value(Scalar(0x00)) }
    }

    bb12: {
        assert(const false, "generator resumed after panicking") -> bb12; // bb12[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
                                         // ty::Const
                                         // + ty: bool
                                         // + val: Value(Scalar(0x00))
                                         // mir::Constant
                                         // + span: /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
                                         // + literal: Const { ty: bool, val: Value(Scalar(0x00)) }
    }

    bb13: {
        unreachable;                     // bb13[0]: scope 0 at /home/jonas/dev/rust/src/test/ui/generator/resume-live-across-yield.rs:21:17: 24:6
    }
}
```

</details>

File: src/librustc_traits/dropck_outlives.rs (hunk 4 above)
Comment (Zoxc) [reply src/librustc_traits/dropck_outlives.rs]: Yeah, I guess we'll need the resume argument here. Extra types should be a conservative option anyway.

**Comment by Zoxc** [comment]
I'd list a test that this won't work:
```rust
#![feature(generators, generator_trait)]

use core::ops::Generator;

fn test(a: impl for<'a> Generator<&'a mut bool>) {}

fn foo() {
    let gen = |arg: &mut bool| {
        yield ();
        *arg = true;
    };
    test(gen);
}
```

**Comment by Zoxc** [comment]
I'd also like a test the initial argument gets dropped by panics.

File: src/librustc_ast_lowering/expr.rs [hunk 6]
```diff
@@ -688,12 +688,12 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
         match generator_kind {
             Some(hir::GeneratorKind::Gen) => {
-                if !decl.inputs.is_empty() {
+                if decl.inputs.len() > 1 {
                     struct_span_err!(
                         self.sess,
                         fn_decl_span,
                         E0628,
-                        "generators cannot have explicit parameters"
+                        "too many parameters for generator (expected 0 or 1 parameters)"
```
Comment (Zoxc) [inline src/librustc_ast_lowering/expr.rs]: Make this: `for a generator`

File: src/test/ui/generator/no-parameters-on-generators.rs [hunk 7]
```diff
@@ -1,8 +1,10 @@
 #![feature(generators)]
 
 fn main() {
-    let gen = |start| { //~ ERROR generators cannot have explicit parameters
```
Comment (Zoxc) [inline src/test/ui/generator/no-parameters-on-generators.rs]: You can just delete this test now.

File: src/librustc_mir/borrow_check/invalidation.rs [hunk 8]
```diff
@@ -170,6 +170,8 @@ impl<'cx, 'tcx> Visitor<'tcx> for InvalidationGenerator<'cx, 'tcx> {
                         self.all_facts.invalidates.push((resume, i));
                     }
                 }
+
+                self.mutate_place(location, resume_arg, Deep, JustWrite);
```
Comment (Zoxc) [inline src/librustc_mir/borrow_check/invalidation.rs]: @matthewjasper Could you check the borrowck changes?

File: src/librustc_mir_build/build/mod.rs [hunk 9]
```diff
@@ -77,22 +83,23 @@ fn mir_build(tcx: TyCtxt<'_>, def_id: DefId) -> BodyAndCache<'_> {
...
                     vec![ArgInfo(liberated_closure_env_ty(tcx, id, body_id), None, None, None)]
                 }
-                ty::Generator(..) => {
+                ty::Generator(def_id, substs, _) => {
                     let gen_ty = tcx.body_tables(body_id).node_type(id);
-                    vec![
-                        ArgInfo(gen_ty, None, None, None),
-                        ArgInfo(tcx.mk_unit(), None, None, None),
-                    ]
+                    let resume_ty = substs.as_generator().resume_ty(def_id, tcx);
```
Comment (Zoxc) [inline src/librustc_mir_build/build/mod.rs]: You should get the `resume_ty` from `gen_ty` as the `ty::Generator` above is supposed to have type parameters for `resume_ty`, `witness`, etc.

File: src/librustc_mir/transform/generator.rs [hunk 10]
```diff
@@ -1110,6 +1128,10 @@ fn create_cases<'tcx>(
                             Rvalue::Use(Operand::Move(resume_arg.into())),
                         )),
                     });
+                    statements.push(Statement {
+                        source_info,
+                        kind: StatementKind::StorageDead(resume_arg),
+                    });
```
Comment (Zoxc) [inline src/librustc_mir/transform/generator.rs]: I don't think you should add a `StorageDead` for the argument here since arguments don't usually use `Storage*` instructions and there's no matching `StorageLive`.

File: src/librustc_mir/transform/generator.rs [hunk 11]
```diff
@@ -1166,7 +1188,28 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
 
         // We rename RETURN_PLACE which has type mir.return_ty to new_ret_local
         // RETURN_PLACE then is a fresh unused local with type ret_ty.
-        let new_ret_local = replace_result_variable(ret_ty, body, tcx);
+        let new_ret_local = replace_local(RETURN_PLACE, ret_ty, body, tcx);
+
+        // We also replace the resume argument and insert an `Assign`.
+        // This is needed because the resume argument might be live across a `yield`, and the
+        // transform assumes that any local live across a `yield` is assigned to before that.
```
Comment (Zoxc) [inline src/librustc_mir/transform/generator.rs]: Where does the transform assume this?

Comment (jonas-schievink) [reply src/librustc_mir/transform/generator.rs]: It doesn't explicitly assume it, but the way it rewrites any mention of locals stored in the generator state doesn't work with an argument that is passed in (and initialized) from the outside. There's never a store to `_2` because it's already initialized, so there won't be any store to its slot in the generator state either, and then the first mention after the yield will read from the uninitialized slot.

File: src/librustc_mir_build/build/mod.rs (hunk 9 above)
Comment (jonas-schievink) [reply src/librustc_mir_build/build/mod.rs]: Done, but I could also just use `tcx.mk_unit()` directly I guess. If you prefer that, I can switch to that.

Comment (Zoxc) [reply src/librustc_mir_build/build/mod.rs]: Why use `()` here? That doesn't seem to relate to the resume type at all.

File: src/librustc_mir/transform/generator.rs (hunk 11 above)
Comment (Zoxc) [reply src/librustc_mir/transform/generator.rs]: I think you should rephrase the comment so it says we need to introduce a store to the resume argument so that such a store to the generator state will exist if it's live across a yield.

File: src/librustc_mir_build/build/mod.rs (hunk 9 above)
Comment (jonas-schievink) [reply src/librustc_mir_build/build/mod.rs]: When the user doesn't provide an argument, the resume type is set to `()`. This code path only has to add an implicit argument if the user didn't write one, so it's always `()` in that case (unless I'm missing something).

Comment (Zoxc) [reply src/librustc_mir_build/build/mod.rs]: Ah yeah, you can just use `()` here.

File: src/librustc_mir/transform/generator.rs (hunk 11 above)
Comment (jonas-schievink) [reply src/librustc_mir/transform/generator.rs]: Yeah, good idea. I've pushed an updated comment.

**Comment by semtexzv** [comment]
@jonas-schievink Thanks for the change, I'd be interested about how it relates to the discussion in https://github.com/rust-lang/rfcs/pull/2781 , I'd love to have your input :)

**Comment by jonas-schievink** [comment]
@semtexzv I don't really have the energy to partake in RFC discussions, but I seem to have implemented the second option described in https://github.com/rust-lang/rfcs/pull/2781#issuecomment-541065199 in this PR

**Comment by pitaj** [comment]
@jonas-schievink as far as I can tell, this PR doesn't introduce any generator syntax for defining the types of resume arguments and yield values. It still relies on type inference to get those types, right?

What would a maximally qualified generator look like now? Something like this?:

    let g = || -> u32 { 5 };

**Comment by jonas-schievink** [comment]
@pitaj The resume argument works like a closure argument, you can give it an explicit type like `|arg: Type|`. From what I can tell, you're right about the yield type though.
